File: applications/bissextile/bissextile.py
# -----------------------------------------------
# Fonction bissextile
# -----------------------------------------------
# Fonction indiquant si une année est bissextile
# en suivant les règles du calendrier grégorien.
# -----------------------------------------------



# -----------------------------------------------
# Version courte
# -----------------------------------------------
# Cette version permet de comprendre le contexte
# de la fonction et sa logique.

def bissextile(annee):
    if annee % 400 == 0:
        return True
    elif annee % 100 == 0:
        return False
    elif annee % 4 == 0:
        return True
    else:
        return False



# -----------------------------------------------
# Application
# -----------------------------------------------

texte = ["n'est pas", "est"]
annees = [1900, 2000, 2024]
for annee in annees:
    est_bis = bissextile(annee)
    chaine = texte[est_bis]
    print(f"{annee} {chaine} bissextile.")



# -----------------------------------------------
# Version golf
# -----------------------------------------------
# Version condensée et optimisée du code, utili-
# sant des noms de variables courts et combinant
# certaines opérations pour réduire la taille du
# code. Pour la beauté du geste !

B=lambda a:a%400==0 or(a%100!=0 and a%4==0)



# -----------------------------------------------
# Version commentée
# -----------------------------------------------
# Similaire à la version détaillée, mais avec des
# commentaires concis afin d'expliquer les étapes
# principales de la fonction.

# On va du plus large vers le plus court.

def bissextile(annee):
    # Si une année est divisible par 400, elle
    # est bissextile.
    if annee % 400 == 0:
        return True
    
    # Sinon, si une année est divisible par 100,
    # sauf 400, elle n'est pas bissextile.
    elif annee % 100 == 0:
        return False
    
    # Sinon, si une année est divisible par 4,
    # elle est bissextile
    elif annee % 4 == 0:
        return True
    
    # Sinon, c'est une année quelconque, enfin,
    # sauf si c'est votre anniversaire.
    else:
        return False









#

File: temps/bissextile/bissextile.js
//-----------------------------------------------
// Fonction bissextile
//-----------------------------------------------
// Fonction indiquant si une année est bissextile
// en suivant les règles du calendrier grégorien.
//-----------------------------------------------

//-----------------------------------------------
// Version courte
//-----------------------------------------------

function bissextile(annee) {
    if (annee % 400 === 0) {
        return true;
    } else if (annee % 100 === 0) {
        return false;
    } else if (annee % 4 === 0) {
        return true;
    } else {
        return false;
    }
}



//-----------------------------------------------
// Application
//-----------------------------------------------

let texte = ["n'est pas", "est"];
let annees = [1900, 2000, 2024];
for (let annee of annees) {

    let est_bis = bissextile(annee);

    let res = annee + " ";

    if (est_bis) {
        res += "est";
    } else {
        res += "n'est pas";
    }
    res += " bissextile.";

    console.log(res);
}



//-----------------------------------------------
// Version golf
//-----------------------------------------------

let B=a=>a%400==0||(a%100!=0&&a%4==0)



//-----------------------------------------------
// Version commentée
//-----------------------------------------------

function bissextile_commentee(annee) {
    // Si une année est divisible par 400, elle
    // est bissextile.
    if (annee % 400 === 0) {
        return true;
    }
    
    // Sinon, si une année est divisible par 100,
    // sauf 400, elle n'est pas bissextile.
    else if (annee % 100 === 0) {
        return false;
    }
    
    // Sinon, si une année est divisible par 4,
    // elle est bissextile
    else if (annee % 4 === 0) {
        return true;
    }
    
    // Sinon, c'est une année quelconque, enfin,
    // sauf si c'est votre anniversaire.
    else {
        return false;
    }
}









//